url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tffst/Tffst.java

```diff
@@ -0,0 +1,1712 @@
+
+	return simpleTffst;
+  }
+
+  /**
+   * returns a Tffst representing this tffst but with simple input labels on each edge. 
+   * it expands TfStrings in simple pairs of TFs adding epsilons
+   * 
+   * @return
+   */
+  public Tffst toSingleInputLabelTransitions() {
+
+	Tffst simpleTffst = new Tffst();
+	HashMap<State, State> m = new HashMap<State, State>();
+
+	State last;
+	Set<State> states = getStates();
+
+	for (State s : states) {
+	  
+	  System.out.println(s.toString());
+	  
+	  if (m.get(s) == null) {
+		last = new State();
+		m.put(s, last);
+		last.accept = s.accept;
+		if (s == initial)
+		  simpleTffst.initial = last;
+	  } else {
+		last = m.get(s);
+	  }
+	  
+	  for (Transition t : s.getTransitions()) {
+
+		if (m.get(t.getTo()) == null) {
+		  State newTo = new State();
+		  m.put(t.getTo(), newTo);
+		  newTo.accept = t.getTo().accept;
+		  if (t.getTo() == initial)
+			  simpleTffst.initial = t.getTo();
+		}
+
+		Transition lastTran = new Transition(last, new TfString(t.labelIn), new TfString(t.labelOut), m.get(t.getTo()));
+
+		while (lastTran.getLabelIn().size() > 1 ) {
+		  last = lastTran.getFrom(); 
+		  State ns = new State();
+		  lastTran.setFrom(ns);
+		  if (lastTran.getLabelIn().size() > 1 ) {
+			new Transition(last, new TfString(lastTran.getLabelIn().remove(0)), new TfString(SimpleTf.Epsilon()), ns);
+		  } 
+		}
+	  }
+	}
+
+	return simpleTffst;
+  }
+
+  
+  
+  /**
+   * 
+   * 
+   * 
+   */
+  public void simplifyTransitionLabels() {
+	Iterator<State> i = getStates().iterator();
+	while (i.hasNext()) {
+	  State s = (State) i.next();
+	  Set<Transition> sTransToRem = new HashSet<Transition>();
+	  Iterator<Transition> j = s.getTransitions().iterator();
+	  while (j.hasNext()) {
+		Transition t = j.next();
+		if (!t.labelIn.isEpsilon())
+		  t.labelIn.set(0, uy.edu.fing.mina.fsa.logics.Utils.simplify(t.labelIn.get(0)));
+		if (t.labelIn.get(0).acceptsNone())
+		  sTransToRem.add(t);
+		else if (!t.labelOut.isEpsilon())
+		  t.labelOut.set(0, uy.edu.fing.mina.fsa.logics.Utils.simplify(t.labelOut.get(0)));
+	  }
+	  s.getTransitions().removeAll(sTransToRem);
+	}
+  }
+  
+  
+  
+  /**
+   * returns a Tffsr representing this tffst. it expands subEpochs in simpe
+   * pairts of TFs adding epsilons
+   * 
+   * @return
+   */
+  public Tffsr toTffsr() {
+
+    Tffst simpleTffst = toSingleLabelTransitions(); 
+    Tffsr outTffsr = new Tffsr();
+    Map<State, uy.edu.fing.mina.fsa.tffsr.State> stateMap = new HashMap<State, uy.edu.fing.mina.fsa.tffsr.State>();
+
+    for (State fstState : simpleTffst.getStates()) {
+      uy.edu.fing.mina.fsa.tffsr.State fsrState = new uy.edu.fing.mina.fsa.tffsr.State();
+      fsrState.setAccept(fstState.isAccept());
+      if (simpleTffst.initial.equals(fstState)) 
+        outTffsr.initial = fsrState;
+      stateMap.put(fstState, fsrState);
+    }
+    
+    for (State fstState : stateMap.keySet()) {
+      uy.edu.fing.mina.fsa.tffsr.State fsrState = stateMap.get(fstState);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.